File: patterns/observer.py
"""Observer example: subject notifies dependents of state changes."""

class Subject:
    def __init__(self):
        self._observers = []
        self._state = None

    def attach(self, observer):
        self._observers.append(observer)

    def set_state(self, value):
        self._state = value
        for observer in self._observers:  # Notify each observer.
            observer.update(value)

class PrintObserver:
    def __init__(self, name):
        self.name = name

    def update(self, value):
        print(f"{self.name} received: {value}")

if __name__ == "__main__":
    subject = Subject()
    subject.attach(PrintObserver("Logger"))
    subject.attach(PrintObserver("Dashboard"))
    subject.set_state("READY")

File: patterns/observer.js
<<<<<<< ours
// Observer example: subject notifies dependents of state changes.
=======
>>>>>>> theirs
class Subject {
  constructor() {
    this._observers = [];
    this._state = null;
  }

  attach(observer) {
    this._observers.push(observer);
  }

  setState(value) {
    this._state = value;
    for (const observer of this._observers) {
<<<<<<< ours
      // Notify each observer.
=======
>>>>>>> theirs
      observer.update(value);
    }
  }
}

class PrintObserver {
  constructor(name) {
    this.name = name;
  }

  update(value) {
    console.log(`${this.name} received: ${value}`);
  }
}

if (require.main === module) {
  const subject = new Subject();
  subject.attach(new PrintObserver("Logger"));
  subject.attach(new PrintObserver("Dashboard"));
  subject.setState("READY");
}

module.exports = { Subject, PrintObserver };
